Add `FnDef` to `TypeName`

nathanwhit opened the pull request:
Addresses the `FnDef` part of #368.

Lowering and program clause generation are probably the areas that need the most review. I'm also a bit unsure about if/how the builtin traits apply to `FnDef`, so some guidance on that would be helpful. For parsing I've chosen to only handle function declarations for simplicity's sake (plus I'm not sure why we would need to inspect the function bodies).

In its current state the PR could use some refactoring/polish, but I'll defer that until the more substantive aspects of the PR are in good shape.

nikomatsakis commented on chalk-solve/src/clauses/program_clauses.rs:
```diff
@@ -337,6 +337,63 @@ impl<I: Interner> ToProgramClauses<I> for StructDatum<I> {
+            let self_ty = self_appl_ty.clone().intern(interner);
+
+            // forall<T> {
+            //     WF(foo<T>) :- WF(T: Eq).
```
This comment kind of makes no sense without a bit more context :)

nikomatsakis replied:
It's intended to refer into an example that would be at the top of the function. For example

```rust
fn foo<T: Eq>() 
```

nikomatsakis commented on chalk-solve/src/clauses/program_clauses.rs:
```diff
@@ -337,6 +337,63 @@ impl<I: Interner> ToProgramClauses<I> for StructDatum<I> {
+                //
+                // In other words, you can assume `T: Bar<&'a i32>`
+                // for any `'a` *if* you are assuming that `foo<T>` is
+                // well formed.
```
Hmm so I guess this code is copy-and-pasted from elsewhere, right? Makes me wonder if we should extract some helper functions here. Seems like it.

nikomatsakis commented on tests/lowering/mod.rs:
```diff
@@ -559,3 +559,28 @@ fn slices() {
+        program {
+            trait Quux { }
+
+            fn foo<'a, T>(bar: T, baz: &'a mut T) -> u32
```
woah. I guess we landed PRs for `&mut T` while I wasn't looking. Exciting. =)

nikomatsakis commented on tests/test/functions.rs:
```diff
@@ -23,3 +23,42 @@ fn functions_are_sized() {
+            WellFormed(garply)
+        } yields {
+            "Unique"
+        }
```
Let's add an implied bounds test. Consider:

```
goal {
    FromEnv(baz<Bar>) => Implemented(Bar: Foo)
}
```

which I think should be provable

nathanwhit commented:
Thanks for the helpful feedback!

I've added the implied bounds test and cleaned up the program clause generation code quite a bit, so it should be ready for re-review.

nikomatsakis commented on chalk-solve/src/clauses/program_clauses.rs:
```diff
@@ -180,6 +180,95 @@ impl<I: Interner> ToProgramClauses<I> for OpaqueTyDatum<I> {
+
+fn implied_bounds_program_clauses<'a, I, Wc>(
+    builder: &'a mut ClauseBuilder<'_, I>,
+    id: impl CastTo<TypeName<I>>,
```
maybe rename to `type_name`?

nikomatsakis commented on chalk-solve/src/clauses/program_clauses.rs:
```diff
@@ -180,6 +180,95 @@ impl<I: Interner> ToProgramClauses<I> for OpaqueTyDatum<I> {
+    );
+}
+
+fn implied_bounds_program_clauses<'a, I, Wc>(
```
This could use a comment. Something like:

/// Generates the "implied bounds" clauses for an applicative 
/// type with the name `type_name`. As an example, if `type_name`
/// represents a struct `S` that is declared like:
///
/// ```rust
/// struct S<T> where T: Eq { .. }
/// ```
///
/// then we would generate the rule
///
/// ```notrust
/// FromEnv(T: Eq) :- FromEnv(S<T>)
/// ```
///
/// # Parameters
///
/// - `builder` -- the clause builder. We assume all the generic types from `S` are in scope.
/// - `type_name` -- in our example above, the name `S`
/// - `where_clauses` -- the list of where clauses declared on the type (`T: Eq`, in our example).

nikomatsakis commented on chalk-solve/src/clauses/program_clauses.rs:
```diff
@@ -180,6 +180,95 @@ impl<I: Interner> ToProgramClauses<I> for OpaqueTyDatum<I> {
+    );
+}
+
+fn fully_visible_program_clauses<'a, I>(
```
Move the comment from inside the function to a doc-comment

nikomatsakis commented on chalk-solve/src/clauses/program_clauses.rs:
```diff
@@ -180,6 +180,95 @@ impl<I: Interner> ToProgramClauses<I> for OpaqueTyDatum<I> {
+    appl_ty.clone().intern(interner)
+}
+
+fn well_formed_program_clauses<'a, I, Wc>(
```
Move comment from inside the function to a doc comment

nikomatsakis commented on chalk-solve/src/clauses/program_clauses.rs:
```diff
@@ -180,6 +180,95 @@ impl<I: Interner> ToProgramClauses<I> for OpaqueTyDatum<I> {
+    }
+}
+
+fn interned_application_ty<I: Interner>(interner: &I, appl_ty: &ApplicationTy<I>) -> Ty<I> {
```
debatable if this helper is worth it, but .. seems fine :)

nikomatsakis commented on chalk-rust-ir/src/lib.rs:
```diff
@@ -103,6 +103,23 @@ pub struct StructFlags {
     pub fundamental: bool,
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
```
Can we add a doc comment explaining what a `FnDef` is, giving some examples

nikomatsakis commented on chalk-rust-ir/src/lib.rs:
```diff
@@ -103,6 +103,23 @@ pub struct StructFlags {
+    pub flags: FnDefFlags,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Fold, HasInterner)]
```
Add doc comment referencing `FnDefDatum`, probably, or maybe giving some more details from an example

nikomatsakis commented on chalk-rust-ir/src/lib.rs:
```diff
@@ -103,6 +103,23 @@ pub struct StructFlags {
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct FnDefFlags {}
```
Hmm an empty set of flags may not be worth including unless we expect to put things in here...

nikomatsakis commented on chalk-rust-ir/src/lib.rs:
```diff
@@ -103,6 +103,23 @@ pub struct StructFlags {
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Fold, HasInterner)]
+pub struct FnDefDatumBound<I: Interner> {
```
I feel like we are going to need .. well, eventually we'll have to add ABI and other details, but maybe we can add that together with adding the same details to the function pointer types

nikomatsakis commented:
r=me but this needs to be rebased.

nikomatsakis commented:
And we should probably file a followup issue regarding ABI details (which apply equally to function pointers)

nathanwhit commented:
Done and done!
